### DISCUSSION TITLE
Publish Metrics using OtlpExporter

### DISCUSSION THREAD

**Discussion by gkatoch** [open-telemetry/opentelemetry-dotnet#3237] [Q&A]
Hi,
I am trying Open telemetry with a very limited knowledge. I am writing a very simple Web API in windows and want to integrate this with OtlpExporter  for publishing metrics. I verified this using console, however I am not able to export the metrics using OLTP exporter.
Sample code used:
```
builder.Services.AddOpenTelemetryMetrics(options =>    
    options.AddOtlpExporter((op1, readOptions) =>
    {
        op1.Protocol = OtlpExportProtocol.HttpProtobuf;
        op1.Endpoint = new Uri("http://localhost:4317/v1/metrics");
        readOptions.PeriodicExportingMetricReaderOptions.ExportIntervalMilliseconds = 5000;

    })    
    .AddHttpClientInstrumentation()
        .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(sourceName).AddTelemetrySdk())
        .AddMeter(meterName));
```
When I run this I have HTTP exceptions for obvious reasons as the endpoint for my exporter is not recognized/reachable. I am not using any docker or any other container for containerization.
How can I expose this metrics to a local HTTP endpoint? 
Do I need to install and setup a collector as from [here](https://github.com/open-telemetry/opentelemetry-collector-releases/releases).
My intention is to setup this endpoint where app will publish data and another exporter can then pull metrics from here. I do not intend to specify Prometheus or other exporter here directly.
Any help will be appreciated.

**Comment by cijothomas** [comment]
> op1.Endpoint = new Uri("http://localhost:4317/v1/metrics"); 

You need to ensure OTel Collector is running and listening to the above port.
See example here https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/examples/Console/TestOtlpExporter.cs

**Reply by gkatoch** [reply]
@cijothomas 
I tried running the Examples.Console Project with arg as: `"otlp"` but this too fails to send the Metrics events in the source example repo itself. Am I missing anything to work with this ? 
`Grpc.Core.RpcException: 'Status(StatusCode="Unavailable", Detail="Error starting gRPC call. HttpRequestException: No connection could be made because the target machine actively refused it. (localhost:4317)`
Is the example not self sufficient to run ?
Thanks

**Reply by cijothomas** [reply]
>  Am I missing anything to work with this ?

Yes. You need to have OTel Collector running.
You may spin one locally in docker, following the instructions from the example.
Or you may follow the instructions from OTel Collector repo on running in any other way.
Or you can send to some vendors, who natively accept OTLP. (eg: New Relic)

**Reply by alanwest** [reply]
Assuming you have docker running on your machine, a comment in the example provides a one-liner (for windows or unix) for getting an instance of the collector running.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/950c6162b43139aeef179bf7e85d70214d02b736/examples/Console/TestOtlpExporter.cs#L29-L52
